### PR TITLE
Add chart component

### PR DESCRIPTION
**Pull request by andysellick**
## What
Adds a chart component.

Based on a component from `content-data-admin` originally added here: https://github.com/alphagov/content-data-admin/pull/9

TODO:

- check the ONS guidance on accessible graphs to see what improvements we can make
- fix the console error when the graph is hovered

## Visual Changes
New component:

Screenshot 2024-10-17 at 10 55 07

Trello card: https://trello.com/c/gvLYQCUp/57-build-statistics-charts

### CODE REVIEW

**Review by AshGDS — COMMENTED** [review]
Code wise it looks good, but I just had a few questions about the mobile styles, as the Axis Labels seem to disappear on mobile, and when you click on a data point, it covers other data points so you can't click on them. Are we OK with this on mobile?

**Comment by AshGDS** [comment]
Also, I appreciate that this was a difficult component to take on, so thanks for doing this 👍

File: app/views/govuk_publishing_components/components/_chart.html.erb [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      <%= render "govuk_publishing_components/components/heading", {
+        text: chart_heading,
+        heading_level: chart_heading_level,
+        margin_bottom: 2,
+      } %>
+    <% end %>
+
+    <div class="gem-c-chart__chart" id="<%= chart_id %>">
+      <div class="gem-c-chart__accessibility-message">
+        This chart is a visual representation of the data available in the <a href="#<%= table_id %>">table</a>.
```
Comment (jon-kirwan) [inline app/views/govuk_publishing_components/components/_chart.html.erb:72]: Should this be added to the locale(s)?

File: app/views/govuk_publishing_components/components/docs/chart.yml [hunk 2]
```diff
@@ -0,0 +1,174 @@
+name: "Chart"
+description: "The chart component presents a data chart and a tabular version of the same data"
+part_of_admin_layout: true
+body: |
+  This component takes a set of data and presents it as a line graph and a table with one or more rows and lines.
+  The chart relies upon chartkick and renders using JavaScript, so the table is provided as a fallback for a lack of JavaScript,
+  an accessible view on the data for screenreaders, and a simple view of the raw data for all users.
+
+  The `overview` option can be used to provide a explanation for screen reader users of what the graph shows.
```
Comment (jon-kirwan) [inline app/views/govuk_publishing_components/components/docs/chart.yml:9]: Sorry really minor one! I think 'provide a explanation' should be 'provide an explanation'.

**Review by jon-kirwan — COMMENTED** [review]
It looks good to me. I've tested across various screens / devices, with VoiceOver, with JS disabled and seems to work well for me. I just added two small comments / questions.

**Comment by andysellick** [comment]
> Code wise it looks good, but I just had a few questions about the mobile styles, as the Axis Labels seem to disappear on mobile, and when you click on a data point, it covers other data points so you can't click on them. Are we OK with this on mobile?

Yeah, the axis labels disappearing as the graph gets narrower is a bit unhelpful, although this seems to be the behaviour of the graph so I think we're stuck with it. I think the problem might be that (certainly with some data) it wouldn't be possible to fit the labels (certainly for the Y axis) in with the graph on mobile - imagine if the Y scale was `0` -> `1,000,000,000`. Hopefully the presence of the tooltips will help. I'm also anticipating that once we start using this component we might find some alternative ways of configuring it to better suit what it gets used for.

On the point about the tooltip, I've just pushed a commit to switch to using HTML tooltips (instead of SVG). It seems now that it's much easier to dismiss the tooltip by tapping elsewhere on the page, so hopefully this has improved things.

File: app/views/govuk_publishing_components/components/_chart.html.erb (hunk 1 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_chart.html.erb]: Yes, thanks good spot. Have added.

File: app/views/govuk_publishing_components/components/docs/chart.yml (hunk 2 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/chart.yml]: Good spot, have fixed 👍

File: app/views/govuk_publishing_components/components/_chart.html.erb [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  rows ||= []
+  keys ||= []
+  chart_overview ||= nil
+  hide_legend ||= false
+
+  chart_id = "chart-id-#{SecureRandom.hex(4)}"
+  table_id = "table-id-#{SecureRandom.hex(4)}"
+
+  require "chartkick"
+  Chartkick.options[:html] = '<div id="%{id}"><noscript><p class="govuk-body">Our charts are built using JavaScript but all the data is also available in the table below.</p></noscript></div>'
```
Comment (AshGDS) [inline app/views/govuk_publishing_components/components/_chart.html.erb:21]: Also, does this text need editorial approval?

**Comment by andysellick** [comment]
> It seems that you can't right click the element - is that intentional?

It's not intentional from us, but appears to be intentional in Google charts, see this stack overflow for info: https://stackoverflow.com/questions/20266832/why-google-charts-are-without-right-click-menu

> The charts seem to have an `aria-label` that is `A chart.` - is the lack of customisation for this intentional? The table has an aria label as well.

This is generated automatically by Google charts. It doesn't seem to be possible to directly customise this text - there are some complicated JS solutions online, which I'm hesitant to implement at this point as I think there's not much more helpful information we can provide at this level - the chart should already have a title, and an accessible description.

Interestingly while digging into this I realised the component has two copies of the table - the one that we create inside the details component, and another one that Google charts automatically creates and appends after the chart. It's visually hidden, so still readable by screenreaders. I'll make a note to see if I can hide it somehow.

> The HTML seems to have `col"` attributes in it, which isn't a valid attribute - is this a chartkick thing we can ignore, or a bug?

I'm not sure, where are these exactly?

> The ONS charts fallback to an image when JS is disabled, so that may be worth noting for the future improvements to this component.

Noted, thanks.

**Comment by AshGDS** [comment]
@andysellick No worries - thanks for looking into those.

I don't think the `col"` thing is actually an issue - instead I think it is a bug with the W3C HTML validator I used on the chart, my bad.

File: app/views/govuk_publishing_components/components/_chart.html.erb (hunk 3 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_chart.html.erb:27]: Possibly. If it's alright with you I think we'll look at that later when we're reviewing the pages as a whole.

File: spec/components/chart_spec.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  end
+
+  it "renders the correct table data horizontally" do
+    data[:table_direction] = "horizontal"
+    render_component(data)
+    assert_select ".govuk-table__body .govuk-table__header:nth-child(1)", text: "2017"
+    assert_select ".govuk-table__cell--numeric", 26
+    assert_select ".govuk-table__header", 14
+    assert_select "td:first", text: "5"
+    assert_select "td:last", text: "121"
```
[the remark this answers is not in the document]
Comment (andysellick) [reply spec/components/chart_spec.rb:52]: Sure, have added.

**Review by AshGDS — APPROVED** [review]
Nice work 👍

**Pull request merged**